src/main.py

```diff
@@ -14,28 +27,54 @@ def is_controller_connected():
 
 
 """
-Script to determine leader and follower roles.
+This function calls the ./bt_toggle script with the specified
+command and address
+"""
+def bt_toggle(command, address):
+    subprocess.call(['./bt_toggle.sh', command, address])
 
-Takes input from the camera to determine if the current vehicle
-should lead or follow based on whether or not there exists an ARTag
-in front of the vehicle.
 
-If an ARTag exists, the vehicle will become a follower vehicle.
-If an ARTag does not exist, the vehicle will become the leader vehicle.
-"""
 def main():
     tag = TagRecognition()
+    bt_address = '5C:BA:37:26:6D:9A'
 
-    # If an ARTag is detected, there is a vehicle ahead of this vehicle so it will follow.
-    # If an ARTag is not detected, no vehicle is present so this vehicle will be a leader.
-    if tag.detect() is not None:
-        follower = Follower()
-        while True:
-            follower.follow()
+    # If an ARTag is detected, the vehicle will try to disconnect from a bluetooth controller and
+    # if it is not connected to one it will follow. If an ARTag is not detected, the vehicle will
+    # will try to connect to the bluetooth controller and if succeful, the vehicle will be a leader.
+    if tag.detect():
+        bt_toggle('disconnect', bt_address)
     else:
-        leader = Leader()
-        while True:
-            leader.lead()
+        bt_toggle('connect', bt_address)
+    if is_controller_connected():
+        vehicle = Leader()
+    else:
+        vehicle = Follower()
+
+    # This loop makes vehicle move. If the vehicle is a leader and sees an ARTag
+    # it will attempt to disconnect from the bluetooth controller, if done
+    # successfully it will become a follower vehicle. If the vehicle is a follower
+    # and does not see an ARTag for more than 5 seconds, the it will try to connect
+    # to the bluetooth controller. If done successfully it will become a leader, otherwise
+    # it will continue as a follower and reset the 5 second timer before trying to
+    # connect to the bluetooth controller again.
+    start_time = time.time()
+    while True:
+        if isinstance(vehicle, Leader):
+            vehicle.lead()
+            if tag.detect():
+                bt_toggle('disconnect', bt_address)
+                if not is_controller_connected():
+                    vehicle = Follower()
+        else:
+            vehicle.follow()
+            if tag.detect():
+                start_time = time.time()
```

Comment:
Why do we start a timer if we see a tag? Shouldn't we be resetting a timer if we don't see a tag?

Reply:
This changes the start time to the current time any time we see a tag, so if a tag is absent for over 5 seconds then the code for connecting to a controller will run.

Reply:
This is backwards. We shouldn't change the start time when we see a tag. We should change the start time when we don't see a tag.

Reply:
The problem is that time.time() only goes upwards so this is the best way to do it without working with threads to make a timer. The point of setting start_time to time.time() is to reset the 5 second counter every time a tag is in frame.